doc/designing_for_trust.md

```diff
@@ -0,0 +1,224 @@
+There is an immediate and obvious problem with this: Rust does not guarantee forward progress and includes diverging control flow that "never returns".
+Thus it is possible for Rust code to never reach certain points in control flow, such as by invoking `panic!()` first.
+Normally, however, `panic!()` will cause "unwinding", which walks back through Rust code to the nearest `catch_unwind`, runningzs `Drop` as it goes.
+
+However, this is not always the case, and `panic!()` may be implemented by other forms of divergence such as immediate termination.
+This may seem surprising, but it is a simple extension of the natural observation that `SIGKILL` exists,
+or its sundry equivalents on non-Unix-like operating systems, and Rust code usually runs under an operating system.
+Rust does not consider terminating Rust code to be a violation of memory safety, because ceasing to progress
+is considered the appropriate way to respond to a situation where the program is not capable of soundly handling further events.
+A possible event that can cause this is the "panic-in-panic" scenario: if unwinding also causes a panic, Rust simply aborts.
+
+In a more targeted fashion, it is possible also to `mem::forget` something with `Drop`, or to wrap it in `ManuallyDrop`.
+Together, these facts mean that a destructor can never be relied on to be run when following arbitrary control flow.
+Only Rust control flow that lacks these features can be expected to run all destructors.
+In other words: `Drop` can be intercepted by both events inside normal Rust code and also "outside" it.
+
+This raises the question: what are destructors "for", if code can't actually rely on them to do anything?
+In Rust, `Drop` arguably primarily exists to dismantle objects so they are not left in invalid states, and to allow reclaiming resources early.
+The first only matters if the invalid state can potentially be reached later in the program. Thus, a destructor is only relevant if the program continues.
+Otherwise, it usually takes unsound code to observe states in something that has been subject to `mem::forget`, as that function consumes the object in question.
+Everything moved into it should be "gone" and no longer visible to the rest of the program state unless `unsafe` has been used.
+The second only matters if the resources are not reclaimed by some other mechanism, such as e.g. process termination.
+
+Thus, if an error inside a function causes Rust to leave behind all of the code that contains broken state,
+then at least hypothetically not running any destructors in that code whatsoever, even if it does not include `mem::forget`, is "not a problem".
+This implies that in response to errors from Postgres, a "correct" (in the sense of "sound, not violating Rust's memory safety properties")
+error handling strategy for PL/Rust is simple abject submission that terminates the function instantly.
```

Comment:
This isn't correct. If you have an object on the stack, Drop is guaranteed to run if the process isn't exited and you don't forget it (or more broadly, move it somewhere into the control of something which may not Drop it). This is required for the pinning invariant to work, as well as for the soundness of numerous patterns in the standard library internals and the broader ecosystem.

IOW, "You cannot rely on Drop because `forget` and `abort` exist" does not mean it is sound for a Rust implementation to fail to destroy an object without running its destructor when the process lives on.

More broadly, what you're suggesting is equivalent (both in theory and in fact in implementation) to forced unwinding across functions with landing pads, and that's just definitely UB.

That said, as of the recent PGX changes, I suspect we shouldn't need to do this anymore? We probably can just use the normal panic runtime?

Reply:
I have definitely learned an unfortunate amount about unwinding to the point I have concluded that I was probably wrong when I initially set out, though I still have to spend some time to wrap my head around the Pin/Unpin nonsense and thus rewrite this section. But yeah, the underlying fixes in PGX can allow us to fix this, and I still think starting from the premise of "we must override GlobalAlloc" was Good, Actually, as it allows us other advantages.

Reply:
> I still think starting from the premise of "we must override GlobalAlloc" was Good, Actually, as it allows us other advantages.

I'm not sure. What about cases where rust code stores data allocated from the `#[global_allocator] in a static and uses it across calls? This is safe if you use a Mutex/RwLock/lazy_static!/OnceCell/... to store it

Won't that be memory-unsafe as the memory will be allocated out of the current context?

Reply:
OH YEAH I'm expecting that we may have to do some seriously silly stuff about `static` involving context control, but that we *can*, because statics **must** be globally unique and thus we can find the damn things, one way or another.